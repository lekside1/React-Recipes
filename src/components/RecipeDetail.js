import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { compose, branch, renderComponent } from 'recompose';


const LoadingState = ({ style, className }) => (
  <p
    style={style}
    className={classNames('h3 p2 bg-gray italic center', className)}
  >
    Loading recipe...
  </p>
);

const ZeroState = ({ style, className }) => (
  <p
    style={style}
    className={classNames('h3 p2 bg-gray italic center', className)}
  >
    No recipe selected
  </p>
);


const RecipeDetail = ({ recipe, style, className }) => {
  const currentLoc = window.location.pathname;

  return (
    <div style={style} className={classNames('p2 bg-gray', className)}>
      <h2 className="h2">{recipe.name}</h2>
      <div className="flex flex-column">
        <img src={recipe.image} alt={recipe.name} className="fit" />
        <div>
          <span>{recipe.category} </span>
          <span>{recipe.calories} cal</span>
        </div>
        {recipe.ingredients && (
          <div>
            <h3>Ingredients</h3>
            <ul>
              {recipe.ingredients.map(ingredient => (
                <li key={ingredient}>{ingredient}</li>
              ))}
            </ul>
          </div>
        )}
        {recipe.steps && (
          <div>
            <h3>Steps</h3>
            <ol>{recipe.steps.map(step => <li key={step}>{step}</li>)}</ol>
          </div>
        )}
        <div>
          {recipe && (
            currentLoc === `/recipe/${recipe.id}`
              ? (
                <Link to="/home" className={classNames('text-decoration-none bg-lime')}>
                  <span role="img" aria-label="less-detail">➖</span>
                </Link>
              )
              : (
                <Link to={`/recipe/${recipe.id}`} className={classNames('text-decoration-none bg-lime')}>
                  <span role="img" aria-label="more-detail">➕</span>
                </Link>
              ))}
        </div>
      </div>
    </div>
  );
};

RecipeDetail.propTypes = {
  recipe: PropTypes.object,
  style: PropTypes.object,
  className: PropTypes.string,
};

RecipeDetail.defaultProps = {
  recipe: {},
  style: {},
  className: '',
};

LoadingState.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
};

LoadingState.defaultProps = {
  style: {},
  className: '',
};

ZeroState.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
};
ZeroState.defaultProps = {
  style: {},
  className: '',
};

export default compose(
  branch(props => props.loading, renderComponent(LoadingState)),
  branch(props => !props.recipe, renderComponent(ZeroState)),
)(RecipeDetail);
