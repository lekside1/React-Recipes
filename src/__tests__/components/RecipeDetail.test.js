import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import RecipeDetail from '../../components/RecipeDetail';

describe('<RecipeDetail />', () => {
  let component;
  let tree;
  let testRecipeDetail;

  beforeEach(() => {
    testRecipeDetail = {
      id: 1,
      name: 'Test Recipe',
      category: 'Dessert',
      calories: 100,
      image: 'test image url',
      ingredients: ['Ingredient 1', 'Ingredient 2'],
      steps: ['Step 1', 'Step 2'],
    };
  });

  const recipe = (props = {}) => (
    <BrowserRouter>
      <RecipeDetail {...props} />
    </BrowserRouter>
  );

  test('Should render RecipeDetail if no recipe', () => {
    component = renderer.create(recipe());
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Should render RecipeDetail recipe', () => {
    component = renderer.create(recipe({ recipe: testRecipeDetail }));
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
