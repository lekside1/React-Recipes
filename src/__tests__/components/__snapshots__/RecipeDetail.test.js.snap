// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeDetail /> Should render RecipeDetail if no recipe 1`] = `
<p
  className="h3 p2 bg-gray italic center"
  style={Object {}}
>
  No recipe selected
</p>
`;

exports[`<RecipeDetail /> Should render RecipeDetail recipe 1`] = `
<div
  className="p2 bg-gray"
  style={Object {}}
>
  <h2
    className="h2"
  >
    Test Recipe
  </h2>
  <div
    className="flex flex-column"
  >
    <img
      alt="Test Recipe"
      className="fit"
      src="test image url"
    />
    <div>
      <span>
        Dessert
         
      </span>
      <span>
        100
         cal
      </span>
    </div>
    <div>
      <h3>
        Ingredients
      </h3>
      <ul>
        <li>
          Ingredient 1
        </li>
        <li>
          Ingredient 2
        </li>
      </ul>
    </div>
    <div>
      <h3>
        Steps
      </h3>
      <ol>
        <li>
          Step 1
        </li>
        <li>
          Step 2
        </li>
      </ol>
    </div>
    <div>
      <a
        className="text-decoration-none bg-lime"
        href="/recipe/1"
        onClick={[Function]}
      >
        <span
          aria-label="more-detail"
          role="img"
        >
          âž•
        </span>
      </a>
    </div>
  </div>
</div>
`;
